File: db4oj/core/src/com/db4o/MReadMultipleObjects.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o;

import com.db4o.foundation.network.YapSocket;

final class MReadMultipleObjects extends MsgD {
	final boolean processMessageAtServer(YapSocket sock) {

		int size = readInt();
		MsgD[] ret = new MsgD[size];
		int length = (1 + size) * YapConst.INT_LENGTH;
		YapStream stream = getStream();

		YapWriter bytes = null;		
		synchronized (stream.i_lock) {
			for (int i = 0; i < size; i++) {
				int id = this._payLoad.readInt();
				try {
					bytes =
						stream.readWriterByID(
							getTransaction(),
							id);
				} catch (Exception e) {
					bytes = null;
				}
				if(bytes != null){
					ret[i] = Msg.OBJECT_TO_CLIENT.getWriter(bytes);
					length += ret[i]._payLoad.getLength();
				}
			}
		}
		
		MsgD multibytes = Msg.READ_MULTIPLE_OBJECTS.getWriterForLength(getTransaction(), length);
		multibytes.writeInt(size);
		for (int i = 0; i < size; i++) {
			if(ret[i] == null){
				multibytes.writeInt(0);
			}else{
				multibytes.writeInt(ret[i]._payLoad.getLength());
				multibytes._payLoad.append(ret[i]._payLoad._buffer);
			}
		}
		multibytes.write(stream, sock);
		return true;
	}
}

File: db4on/core/src/core/com/db4o/cs/messages/MReadMultipleObjects.cs
namespace com.db4o.cs.messages
{
	public sealed class MReadMultipleObjects : com.db4o.cs.messages.MsgD
	{
		public sealed override bool ProcessMessageAtServer(com.db4o.foundation.network.YapSocket
			 sock)
		{
			int size = ReadInt();
			com.db4o.cs.messages.MsgD[] ret = new com.db4o.cs.messages.MsgD[size];
			int length = (1 + size) * com.db4o.YapConst.INT_LENGTH;
			com.db4o.YapStream stream = GetStream();
			com.db4o.YapWriter bytes = null;
			lock (stream.i_lock)
			{
				for (int i = 0; i < size; i++)
				{
					int id = this._payLoad.ReadInt();
					try
					{
						bytes = stream.ReadWriterByID(GetTransaction(), id);
					}
					catch
					{
						bytes = null;
					}
					if (bytes != null)
					{
						ret[i] = com.db4o.cs.messages.Msg.OBJECT_TO_CLIENT.GetWriter(bytes);
						length += ret[i]._payLoad.GetLength();
					}
				}
			}
			com.db4o.cs.messages.MsgD multibytes = com.db4o.cs.messages.Msg.READ_MULTIPLE_OBJECTS
				.GetWriterForLength(GetTransaction(), length);
			multibytes.WriteInt(size);
			for (int i = 0; i < size; i++)
			{
				if (ret[i] == null)
				{
					multibytes.WriteInt(0);
				}
				else
				{
					multibytes.WriteInt(ret[i]._payLoad.GetLength());
					multibytes._payLoad.Append(ret[i]._payLoad._buffer);
				}
			}
			multibytes.Write(stream, sock);
			return true;
		}
	}
}
